src/vm/object_model.rs

```diff
@@ -7,6 +7,50 @@ use crate::util::metadata::MetadataValue;
 use crate::util::{Address, ObjectReference};
 use crate::vm::VMBinding;
 
+/// The strategy to update the valid object (VO) bits metadata for ImmixSpace.
+///
+/// It only affects plans that use ImmixSpace, including Immix, GenImmix and StickyImmix.
+///
+/// Note that some strategies have implications on the availability of VO bits and the layout of
+/// the mark bits metadata.  VM bindings should choose the appropriate strategy according to its
+/// specific needs.
+///
+/// See also: [`crate::util::metadata::vo_bit`]
+#[cfg(feature = "vo_bit")]
+#[derive(Debug)]
+pub enum ImmixVOBitUpdateStrategy {
```

Comment:
> During tracing (when we are in the process of identifying object validity),

There is a difference between validity and reachability.

-   An object reference is **valid** if it points to an object that has been allocated by MMTk.
-   During tracing, we compute the **reachability** of objects, i.e. whether we can reach it from root.
-   At the end of GC, unreachable objects are no longer valid.

> The VO bits is always imprecise during tracing.

From the user's point of view, the user of VO bits use it for deciding whether a conservative edge (from root or from field) is valid. By "valid", it means if it points to a valid object **at the moment when GC started**.  In this sense, I don't think the VO bits metadata is imprecise during tracing.  During tracing, every valid edge should point to an object in the from-space (and for VMs that may visit an edge twice, the GC may see an edge pointing to an object in the to-space, too, as I mentioned in another comment).  Every invalid edge, on the other hand, do not point to any object in either from- or to-space.

If, during tracing, the GC visits an object that has a conservative field, and the VO bit is 1 for the address it contains, it means that field may have pointed to an object before the GC happened, and that address is in the from space of the current GC.  My previous work about "potential pinning parent (PPP)" ensures that such an object must have been pinned (but may not have been marked from another edge).  Assume the object is not marked, then it is **valid** but **unreachable** at this moment.  We mark it by tracing this conservative edge, and it becomes **reachable** while remaining **valid**.

As for the similarity and difference between the two strategies, both strategies guarantee that **before and after GC**, an ObjectReference with VO bit being 1 is definitely valid, and an ObjectReference with VO bit being 0 is definitely invalid.  The difference is during tracing (specifically during the `Closure` and `RefClosure` stage.

-   For `ClearAndReconstruct`, the VO bits are simply not available during tracing;
-   For `CopyFromMarkBits`, the VO bits are set for all from-space objects and all to-space objects.

I agree that the semantics need to precise to the user, but I am reluctant to rename the strategies based on semantics.  Each strategy has its limitation, and there exist other strategies that provide similar semantics, implemented differently, but do not have the same limitations.  For example, `CopyFromMarkBits` requires on-the-side mark bits.  There is another strategy that keeps two copies of VO bits which makes the VO bits available during tracing, is more costly than `CopyFromMarkBits`, but does not require on-the-side mark bits.  The VM needs to choose one strategy according to its own requirement.

Reply:
> As for the similarity and difference between the two strategies, both strategies guarantee that **before and after GC**, an ObjectReference with VO bit being 1 is definitely valid, and an ObjectReference with VO bit being 0 is definitely invalid. The difference is during tracing (specifically during the `Closure` and `RefClosure` stage.
> 
> * For `ClearAndReconstruct`, the VO bits are simply not available during tracing;
> * For `CopyFromMarkBits`, the VO bits are set for all from-space objects and all to-space objects.

Yeah. This is clear. 
 
> I agree that the semantics need to precise to the user, but I am reluctant to rename the strategies based on semantics. Each strategy has its limitation, and there exist other strategies that provide similar semantics, implemented differently, but do not have the same limitations. For example, `CopyFromMarkBits` requires on-the-side mark bits. There is another strategy that keeps two copies of VO bits which makes the VO bits available during tracing, is more costly than `CopyFromMarkBits`, but does not require on-the-side mark bits. The VM needs to choose one strategy according to its own requirement.

How about using `const MAKE_VO_BITS_AVAILABLE_IN_TRACING: bool = false` in `ObjectModel` (or something like this)? Based on the value, internally you can choose which strategy to use, and all the internal code depends on the strategy. My intention is that we would like to make our public interface easy to understand for binding implementers, who care about what semantics we deliver rather than how we deliver it. If I were a binding implementer without much knowledge about MMTk core, I would be very confused if I had to choose between `ClearAndReconstruct` and `CopyFromMarkBits`.

Reply:
> * For `ClearAndReconstruct`, the VO bits are simply not available during tracing;

So for this option, we do not need to do anything special for VO bits, right? It does not guarantee anything during tracing. For different policies, we can maintain vo bits, clear vo bits, or whatever. We can just use the most efficient implementation for a policy, and we only need to make sure that VO bits are correct before and after GC.

Reply:
> How about using `const MAKE_VO_BITS_AVAILABLE_IN_TRACING: bool = false` in `ObjectModel` (or something like this)? Based on the value, internally you can choose which strategy to use, and all the internal code depends on the strategy. My intention is that we would like to make our public interface easy to understand for binding implementers, who care about what semantics we deliver rather than how we deliver it. If I were a binding implementer without much knowledge about MMTk core, I would be very confused if I had to choose between `ClearAndReconstruct` and `CopyFromMarkBits`.

I think that's OK.  If we expose the presence of one strategy, we may have trouble if we no longer want to support that strategy in the future but the VM depends on lt.

My primary concern is that each strategy has its limitation, and the VM has to choose carefully if it needs VO bits.

Let's consider three strategies instead of two, with `ReplicatedVOBits` as the third strategy that keeps the old copy of VO bits during tracing, but constructs the new VO bits in the new copy and copy it over to the old VO bits.

| strategy | ClearAndReconstruct | CopyFromMarkBits | ReplicatedVOBits |
|---|---|---|---|
| available during tracing | no | yes | yes |
| requires on-the-side mark bits | no | yes | no |
| time overhead | low | high | high |
| space overhead | low | low | high |

If a VM needs VO bits, it must need it for some reasons.  If it does, I think the VM binding developer knows exactly what their limitations are, and want more precise control over the strategy it uses.  For example, if the VM uses in-header mark bits, it may not choose the CopyFromMarkBits strategy.

Currently the requirement of on-the-side mark bits can be ignored because we only support on-the-side mark bits at this moment.  Given that we only have two strategies for the user to choose from, I think it is OK to expose only a Boolean variable as option.

Reply:
I removed the `VOBitUpdateStrategy` from the API, and replaced the constant with a `const NEED_VO_BITS_DURING_TRACING: bool = false;`